pkg/devserver/discovery.go

```diff
@@ -0,0 +1,105 @@
+package devserver
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+)
+
+var (
+	// Ports indicate the default ports that we attempt to scan on localhost
+	// when discovering SDK-based endpoints
+	Ports = []int{80, 3000, 3001, 3002, 3003, 3004, 3005, 5000, 8000, 8080, 8081, 8888}
+
+	// Paths indicate the paths we attempt to hit when a web server is available.
+	// These are the default, recommended paths for hosting Inngest routes.
+	Paths = []string{"/api/inngest", "/x/inngest", "/.netlify/functions/inngest"}
+
+	timeout = 2 * time.Second
+
+	hc = http.Client{
+		Timeout: timeout,
+	}
+)
+
+// Autodiscover attempts to automatically discover SDK endpoints running on
+// the local machine, using a combination of the default supported ports
+// and default API endpoints above.
+func Autodiscover(ctx context.Context) []string {
+	results := []string{}
+	ports := openPorts(ctx)
+	for _, port := range ports {
+		for _, path := range Paths {
+			// These requests _should_ be fast as we know a port is open,
+			// so we do these sequentially.
+			url := fmt.Sprintf("http://127.0.0.1:%d%s", port, path)
+			if err := checkURL(ctx, url); err == nil {
+				results = append(results, url)
+			}
+		}
+	}
+	return results
+}
+
+// checkURL attempts to discover whether there's an SDK hosted at the given url
+// by checking the presence of the "x-inngest-sdk" header.
+func checkURL(ctx context.Context, url string) error {
+	resp, err := hc.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.Header.Get("x-inngest-sdk") != "" {
+		// This is valid.
+		return nil
+	}
+	return fmt.Errorf("SDK header not found")
+}
+
+// openPorts simultaneously checks all supported localhost ports to see if
+// any are open. This allows us to filter the default ports to only those
+// that are serving connections prior to making HTTP requests.
+func openPorts(ctx context.Context) []int {
+	results := []int{}
+	// Create a buffered channel with the number of ports, letting us push
+	// valid ports without reading from the channel at the same time.
+	found := make(chan int, len(Ports))
```

Comment:
Should we also check for a `PORT` environment variable? That could be an enhancement in future versions.

Reply:
In this case I think you'd use `--sdk-url` / `-u` to specify the canonical URL you'd like to hit (which is usually 127.0.0.1:whatever!) :)

Reply:
Yeah - I was just thinking for the sake of auto-discovery it would be a cool automatic setup. If we extended #218 to also make environment variables available in the Dev Server code, devs who use Next.js and set `PORT` would just be able to run `inngest dev` in their project directory and it would just pick it up automatically not having to specify the URL arg. With how smart the auto-discovery already is, it could be a nice addition - not blocking merging this PR.

Reply:
Added: https://github.com/inngest/inngest/pull/259/files#diff-f33add88ab91a28b9194b39a63ae00da34d0c7928432bf7aa0d8eec7cb0a9eb4R30-R36.  Thanks1